### PR TITLE
Add support for running macOS guests.

### LINKED ISSUE #139: Running macOS guests
It looks like the most common use of vfkit is running GNU/Linux guests, but of course the Virtualization framework also supports running macOS guests and vz can [download and run macOS guests from .ipsw files](https://github.com/Code-Hex/vz/blob/main/virtualization_arm64.go#L401).  Is it possible to run a macOS guest using vfkit (ideally from a .ipsw)?

### PR DESCRIPTION
**Pull request by williamtheaker**
This PR implements running macOS guests with vfkit, which requires an arm64/Apple silicon device running macOS 12 or later. Thank you cfergeau for your assistance in developing this feature.

## Core concepts

The [macOS in QEMU (ARM edition)](https://www.youtube.com/watch?v=oZqFYJVOUQo) talk explains this in more detail, but booting a macOS guest image requires three components:
* An auxiliary storage file with NVRAM contents and the iBoot bootloader
* hardwareModel - A binary property list defining OS version support
* MachineIdentifier - unique identifier for the VM

These files can be created with Apple's sample [InstallationTool](https://developer.apple.com/documentation/virtualization/running_macos_in_a_virtual_machine_on_apple_silicon) code or by building and running the [installation example](https://github.com/Code-Hex/vz/tree/main/example/macOS) in the vz repo.

## Tested working

Host running 14.5 (23F79) and guest running 14.4.1 (23E224):

* Networking
* Input. I didn't implement VZMacKeyboardConfiguration (macOS 14+) or VZMacTrackpadConfiguration (macOS 13+) since the basic virtio input works fine and I don't need gestures.
* SSH.  Enabling it in the guest via System Preferences and grab the VM ip from the host `/var/db/dhcpd_leases`.
* Adding a virtio-serial device creates a file at `/dev/tty.virtio`.
* RESTful API. It seems like #67 impacts macOS guests too.
* Sharing directories from host to guest

### Unimplemented

* Downloading .ipsw restore images and creating new base install images
* Audio, since vfkit hasn't implemented it yet.
* Various features requiring macOS version 14+:
   * [saveMachineStateTo](https://developer.apple.com/documentation/virtualization/vzvirtualmachine/4168516-savemachinestateto)
   * [restoreMachineStateFrom](https://developer.apple.com/documentation/virtualization/vzvirtualmachine/4168515-restoremachinestatefrom)
* Testing macOS vms similar to the existing puipui-linux tests

## Todo

Figure out how to handle metadata. Other projects like [macosvm](https://github.com/s-u/macosvm?tab=readme-ov-file#details) and Tart create .json files with the VM's hardwareModel and MachineIdentifier values. Maybe this would be a good use for https://github.com/crc-org/vfkit/issues/28 too?

Closes #139

### CODE REVIEW

**Comment by cfergeau** [comment]
> Handle x86 so the Makefile doesn't need to be partially commented out.

I've fixed this in https://github.com/cfergeau/vfkit/tree/wt.macos_guests

**Comment by cfergeau** [comment]
> - Figure out how to handle metadata. Other projects like [macosvm](https://github.com/s-u/macosvm?tab=readme-ov-file#details) and Tart create .json files with the VM's hardwareModel and MachineIdentifier values. Maybe this would be a good use for https://github.com/crc-org/vfkit/issues/28 too?

Do you have an example of the json file used by macosvm or tart?

> - Fix hardcoded values AuxiliaryStorageVar, HardwareModelVar, and MachineIdentifierVar in pkg/vf/vm.go. I could use your help @cfergeau figuring out how to pass the macOS bootloader arguments to NewMacPlatformConfiguration().

I'll try to look at this today!

**Comment by williamtheaker** [comment]
> Do you have an example of the json file used by macosvm or tart?

macosvm
```json
{
  "serial": true,
  "os": "macos",
  "hardwareModel": "YnBsaXN0MDDTAQIDBAUGXxAZRGF0YVJlcHJlc2VudGF0aW9uVmVyc2lvbl8QD1BsYXRmb3JtVmVyc2lvbl8QEk1pbmltdW1TdXBwb3J0ZWRPUxQAAAAAAAAAAAAAAAAAAAABEAKjBwgIEA0QAAgPKz1SY2VpawAAAAAAAAEBAAAAAAAAAAkAAAAAAAAAAAAAAAAAAABt",
  "storage": [
    {
      "type": "disk",
      "file": "disk.img"
    },
    {
      "type": "aux",
      "file": "aux.img"
    }
  ],
  "ram": 4294967296,
  "machineId": "YnBsaXN0MDDRAQJURUNJRBMlAGfNcIpmowgLEAAAAAAAAAEBAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAAZ",
  "displays": [
    {
      "dpi": 200,
      "height": 1600,
      "width": 2560
    }
  ],
  "version": 1,
  "cpus": 2,
  "networks": [
    {
      "type": "nat"
    }
  ],
  "audio": false
}
```

Tart
```json
{
  "macAddress" : "c2:0e:15:c5:6e:25",
  "memorySizeMin" : 4294967296,
  "display" : {
    "width" : 1024,
    "height" : 768
  },
  "hardwareModel" : "YnBsaXN0MDDTAQIDBAUGXxAZRGF0YVJlcHJlc2VudGF0aW9uVmVyc2lvbl8QD1BsYXRmb3JtVmVyc2lvbl8QEk1pbmltdW1TdXBwb3J0ZWRPUxQAAAAAAAAAAAAAAAAAAAABEAKjBwgIEA0QAAgPKz1SY2VpawAAAAAAAAEBAAAAAAAAAAkAAAAAAAAAAAAAAAAAAABt",
  "cpuCount" : 4,
  "os" : "darwin",
  "ecid" : "YnBsaXN0MDDRAQJURUNJRBQAAAAAAAAAAMSGIoemH\/\/VCAsQAAAAAAAAAQEAAAAAAAAAAwAAAAAAAAAAAAAAAAAAACE=",
  "version" : 1,
  "memorySize" : 4294967296,
  "arch" : "arm64",
  "cpuCountMin" : 2
}
```

You can see the hardwareModel is identical in both files. It's a binary plist containing `{"DataRepresentationVersion":1,"MinimumSupportedOS":[13,0,0],"PlatformVersion":2}`. I need to see if I can use that with `vz.NewMacMachineIdentifierWithData` as a fallback.

I'm still trying to figure out whether it makes sense to add the ability to create/install macOS VMs with vfkit, which will influence how to provide these values, whether as strings or files on disk.

**Comment by williamtheaker** [comment]
> > Handle x86 so the Makefile doesn't need to be partially commented out.
> 
> I've fixed this in https://github.com/cfergeau/vfkit/tree/wt.macos_guests Only tried to compile, did not test runtime behaviour

Thanks! Should I merge your architecture-specific files commit into my branch?

**Comment by cfergeau** [comment]
> > > Handle x86 so the Makefile doesn't need to be partially commented out.
> > 
> > 
> > I've fixed this in https://github.com/cfergeau/vfkit/tree/wt.macos_guests Only tried to compile, did not test runtime behaviour
> 
> Thanks! Should I merge your architecture-specific files commit into my branch?

Sure, you can do whatever you see fit with them, cherry-pick, squash, rewrite, ... :)

**Comment by cfergeau** [comment]
> Fix hardcoded values AuxiliaryStorageVar, HardwareModelVar, and MachineIdentifierVar in pkg/vf/vm.go. I could use your help @cfergeau figuring out how to pass the macOS bootloader arguments to NewMacPlatformConfiguration().

I added a few more commits to https://github.com/cfergeau/vfkit/tree/wt.macos_guests to pass the args to NewMacPlatformConfiguration.
There seems to be a bigger question though, where should these values be coming from, who creates them, ..., do they need to be stored in a directory in `~/Library/Application\ Support/vfkit`, ... 
I need to read more about macos VMs to be able to give meaningful advice about this.

> I'm still trying to figure out whether it makes sense to add the ability to create/install macOS VMs with vfkit, which will influence how to provide these values, whether as strings or files on disk.

Same answer as in the previous paragraph, I'll need to look closer at this. My gut feeling is that it would make sense to offer this, it's (a bit) similar to https://github.com/crc-org/vfkit/issues/124 / https://github.com/crc-org/vfkit/issues/123

**Comment by cfergeau** [comment]
If the code is this PR is enough to be able to start and use an already installed macos VM, I think we should work on merging this first. The other missing features can be added on top of this in separate PRs.

**Comment by williamtheaker** [comment]
I merged in your commits and fixed some linting issues. I think this is ready to be reviewed.

**Comment by williamtheaker** [comment]
There's another tool using the Virtualization Framework called Orka, which, like Tart, stores images and "layers" (not sure how closely they match Linux container functionality) in an OCI format. The software is proprietary and it uses nonstandard mediatypes, but we can still look at their config file layout. Here's the config from this [Orka base image](https://github.com/macstadium/orka-images/pkgs/container/orka-images%2Fventura/207398699?tag=no-sip):

```json
{
  "UID": "882B4402-E965-43E4-9570-EA01EEDCE5AA",
  "cpuCount": 4,
  "diskSizeGB": 90,
  "displayDPI": 96,
  "displayHeight": 1080,
  "displayWidth": 1920,
  "graphicalConsole": false,
  "hardwareModel": "YnBsaXN0MDDTAQIDBAUGXxAZRGF0YVJlcHJlc2VudGF0aW9uVmVyc2lvbl8QD1BsYXRmb3JtVmVyc2lvbl8QEk1pbmltdW1TdXBwb3J0ZWRPUxQAAAAAAAAAAAAAAAAAAAABEAKjBwgIEA0QAAgPKz1SY2VpawAAAAAAAAEBAAAAAAAAAAkAAAAAAAAAAAAAAAAAAABt",
  "installed": true,
  "machineIdentifier": "YnBsaXN0MDDRAQJURUNJRBMT1iLyrVR8AggLEAAAAAAAAAEBAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAAZ",
  "memorySize": 4294967296,
  "memorySizeMB": 4096,
  "restoreImage": "default",
  "serialConsole": true,
  "version": "8C89B5C1-3D6E-4B80-A45B-51EE97B6C0EB"
}
```

Looks like everyone is using the same generic hardwareModel string 😅

**Comment by cfergeau** [comment]
This hardware model string is not really special/magic:
```
$ echo YnBsaXN0MDDTAQIDBAUGXxAZRGF0YVJlcHJlc2VudGF0aW9uVmVyc2lvbl8QD1BsYXRmb3JtVmVyc2lvbl8QEk1pbmltdW1TdXBwb3J0ZWRPUxQAAAAAAAAAAAAAAAAAAAABEAKjBwgIEA0QAAgPKz1SY2VpawAAAAAAAAEBAAAAAAAAAAkAAAAAAAAAAAAAAAAAAABt |base64 -d |plistutil -i -

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>DataRepresentationVersion</key>
	<integer>1</integer>
	<key>PlatformVersion</key>
	<integer>2</integer>
	<key>MinimumSupportedOS</key>
	<array>
		<integer>13</integer>
		<integer>0</integer>
		<integer>0</integer>
	</array>
</dict>
</plist>
```

However, I guess Apple does not document it, so better to treat it as an opaque blob.

**Comment by williamtheaker** [comment]
There's a comment in `pkg/vf/vm_arm64.go` with the plist payload. Since the value seems to be unchanged since 2022, maybe in a future commit we can default to using it as a hardcoded value when no hardwareModel argument is supplied.

File: README.md [hunk 1]
```diff
@@ -17,6 +17,9 @@ brew tap cfergeau/crc
 brew install vfkit
 ```
 
+### Building
+
+From the root direction of this repository, run `make`. You will need a valid Apple Developer certificate to sign the binaries with the required entitlements.
```
Comment (cfergeau) [inline README.md]: A valid Apple Developer certificate is not required to sign the binary, the Makefile uses `codesign -s -` aka ad hoc signing ( https://ss64.com/mac/codesign.html ), which works without any Apple certificate.

File: doc/quickstart.md [hunk 2]
```diff
@@ -5,14 +5,14 @@
 vfkit is a macOS command-line-based hypervisor, which uses [Apple's Virtualization Framework](https://developer.apple.com/documentation/virtualization?language=objc) to run virtual machines.
 You start a virtual machine by running vfkit with a set of arguments describing the virtual machine configuration/hardware.
 When vfkit stops, the virtual machine stops running.
-It requires macOS 11 or newer, and runs on both x86_64 and aarch64 Macs.
+It requires macOS 11 or newer, and runs on both x86_64 and arm64 Macs.
```
Comment (cfergeau) [inline doc/quickstart.md]: I usually use either `amd64`/`arm64` which is the go naming, or `x86_64`/`aarch64` which is the linux naming, `x86_64`+`arm64` is imo weirdly inconsistent. We can use `Intel and Apple silicon Macs` which is the naming used by Apple.

File: doc/usage.md [hunk 3]
```diff
@@ -385,34 +402,45 @@ None
...
 
 ### Inspect VM
 
 Get description of the virtual machine
 
-GET `/vm/inspect`
+```HTTP
+GET /vm/inspect
+```
+
```
Comment (cfergeau) [inline doc/usage.md:444]: Thanks for all the typo/formatting/... fixes, I've split them to their own commit.

File: pkg/vf/virtio.go [hunk 4]
```diff
@@ -328,6 +340,7 @@ func (dev *USBMassStorage) AddToVirtualMachineConfig(vmConfig *VirtualMachineCon
 	return nil
 }
 
+// Move these to top of file?
```
Comment (cfergeau) [inline pkg/vf/virtio.go]: I've removed this comment, but yeah it would make sense to group all hte public structs at the top of the file.

File: pkg/vf/vm_arm64.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	"fmt"
+
+	"github.com/Code-Hex/vz/v3"
+	"github.com/crc-org/vfkit/pkg/config"
+)
+
+func NewMacPlatformConfiguration(machineIdentifierVar, hardwareModelVar, auxiliaryStorageVar string) (vz.PlatformConfiguration, error) {
+	// var HardwareModelVar =[]byte( "YnBsaXN0MDDTAQIDBAUGXxAZRGF0YVJlcHJlc2VudGF0aW9uVmVyc2lvbl8QD1BsYXRmb3JtVmVyc2lvbl8QEk1pbmltdW1TdXBwb3J0ZWRPUxQAAAAAAAAAAAAAAAAAAAABEAKjBwgIEA0QAAgPKz1SY2VpawAAAAAAAAEBAAAAAAAAAAkAAAAAAAAAAAAAAAAAAABt") // Binary plist with {"DataRepresentationVersion":1,"MinimumSupportedOS":[13,0,0],"PlatformVersion":2}
```
Comment (cfergeau) [inline pkg/vf/vm_arm64.go]: I've kept the comment, but explained a bit more what this is about:
```
        // The following string is common for the hardware model:
        // `YnBsaXN0MDDTAQIDBAUGXxAZRGF0YVJlcHJlc2VudGF0aW9uVmVyc2lvbl8QD1BsYXRmb3JtVmVyc2lvbl8QEk1pbmltdW1TdXBwb3J0ZWRPUxQAAAAAAAAAAAAAAAAAAAABEAKjBwgIEA0QAAgPKz1SY2VpawAAAAAAAAEBAAAAAAAAAAkAAAAAAAAAAAAAAAAAAABt`
        // It is a base64-encoded binary plist with this content: `{"DataRepresentationVersion":1,"MinimumSupportedOS":[13,0,0],"PlatformVersion":2}`
```

File: cmd/vfkit/main.go [hunk 6]
```diff
@@ -172,7 +172,8 @@ func runVirtualMachine(vmConfig *config.VirtualMachine, vm *vf.VirtualMachine) e
 		log.Debugf("%v", err)
 	}
 
-	log.Infof("waiting for VM to stop")
+	// Is this print statement necessary? It gives the impression that any running VM is in the process of stopping
+	log.Infof("Waiting for VM to stop")
```
Comment (cfergeau) [inline cmd/vfkit/main.go]: I've dropped these changes for now, but I agree this could be improved.

**Review by cfergeau — APPROVED** [review]
Finally got back to this PR, I squashed/split the commits a bit, and dropped the comments which were asking questions about potential fixes.
I've force-pushed these changes to this PR, I'm fine to merge these as they are.
